src/Main.elm

```diff
@@ -1,71 +1,392 @@
+    | OnRename
+
+
+init : () -> ( Model, Cmd Msg )
+init _ =
+    ( { newTodoTitle = ""
+      , todos = []
+      , hideCompleted = False
+      , todoBeingEdited = -1
+      }
+    , fetchTodos
+    )
+
+
 
 -- UPDATE
 
 
-update : msg -> Model -> Model
+fetchTodos : Cmd Msg
+fetchTodos =
+    get "https://todo-backend-hanami.herokuapp.com/"
+        |> withExpect (Http.expectJson LoadedTodos todoListDecoder)
+        |> request
+
+
+postTodo : String -> Cmd Msg
+postTodo title =
+    post "https://todo-backend-hanami.herokuapp.com/"
+        |> withBody
+            (Http.jsonBody
+                (Encode.object
+                    [ ( "title", Encode.string title ) ]
+                )
+            )
+        |> withExpect (Http.expectJson AddedTodo todoDecoder)
+        |> request
+
+
+updateTodo : Todo -> Bool -> String -> Cmd Msg
+updateTodo todo isCompleted title =
+    patch todo.url
+        |> withBody
+            (Http.jsonBody
+                (Encode.object
+                    [ ( "completed", Encode.bool isCompleted )
+                    , ( "title", Encode.string title )
+                    ]
+                )
+            )
+        |> withExpect (Http.expectJson ChangedCompleted todoDecoder)
+        |> request
+
+
+deleteTodo : Todo -> Cmd Msg
+deleteTodo todo =
+    delete todo.url
+        |> withExpect (Http.expectWhatever DeletedTodo)
+        |> request
+
+
+focusTodoEdit : Cmd Msg
+focusTodoEdit =
+    Task.attempt (\_ -> NoOp) (Dom.focus "todo-edit-input")
+
+
+addTodo : Model -> ( Model, Cmd Msg )
+addTodo model =
+    ( { model | newTodoTitle = "" }, postTodo model.newTodoTitle )
+
+
+type Msg
+    = AddTodo
+    | ChangeInput String
+    | EditTodo Todo String
+    | StartEditing Todo
+    | EndEditing Todo
+    | ChangeCompleted Todo Bool
+    | KeyDown KeyDownWhere Int
+    | DeleteTodo Todo
+    | HideCompleted Bool
+    | LoadedTodos (Result Http.Error (List Todo))
+    | AddedTodo (Result Http.Error Todo)
+    | ChangedCompleted (Result Http.Error Todo)
+    | DeletedTodo (Result Http.Error ())
+    | NoOp
+
+
+update : Msg -> Model -> ( Model, Cmd Msg )
 update msg model =
     case msg of
-        _ ->
-            model
+        AddTodo ->
+            addTodo model
+
+        ChangeInput title ->
+            ( { model | newTodoTitle = title }, Cmd.none )
+
+        ChangeCompleted theTodo checked ->
+            ( { model
+                | todos =
+                    List.map
+                        (\todo ->
+                            if todo.id == theTodo.id then
+                                { todo | completed = checked }
+
+                            else
+                                todo
+                        )
+                        model.todos
+              }
+            , updateTodo theTodo checked theTodo.title
+            )
+
+        StartEditing todo ->
+            ( { model | todoBeingEdited = todo.id }, focusTodoEdit )
+
+        KeyDown OnAdd keyCode ->
```

Comment:
Can you pattern match on the value of the `keyCode`? to avoid the if

Reply:
Yes, that seems to work fine, thanks